fil_fungible_token/src/token/mod.rs

```diff
@@ -32,39 +32,61 @@ where
 {
     /// Injected blockstore. The blockstore must reference the same underlying storage under Clone
     bs: BS,
-    /// Root of the token state tree
-    state_cid: Cid,
+    /// In-memory cache of the state tree
+    state: TokenState,
 }
 
 impl<BS> Token<BS>
 where
     BS: IpldStore + Clone,
 {
-    /// Instantiate an empty token helper with access to a blockstore and runtime
-    pub fn new(bs: BS, state_cid: Cid) -> Self {
-        Self { bs, state_cid }
+    /// Creates a new token instance using the given blockstore and creates a new empty state tree
+    ///
+    /// Returns a Token handle that can be used to interact with the token state tree and the Cid
+    /// of the state tree root
+    pub fn new(bs: BS) -> Result<(Self, Cid)> {
+        let init_state = TokenState::new(&bs)?;
+        let cid = init_state.save(&bs)?;
+        let token = Self {
+            bs,
+            state: init_state,
+        };
+        Ok((token, cid))
     }
 
-    /// Constructs the token state tree and saves it at a CID
-    pub fn init_state(&mut self) -> Result<Cid> {
-        let init_state = TokenState::new(&self.bs)?;
-        self.save_state(&init_state)
+    /// For an already initialised state tree, loads the state tree from the blockstore and returns
+    /// a Token handle to interact with it
+    pub fn load(bs: BS, state_cid: Cid) -> Result<Self> {
+        let state = TokenState::load(&bs, &state_cid)?;
+        Ok(Self { bs, state })
     }
 
-    /// Helper function that loads the root of the state tree related to token-accounting
-    ///
-    /// Actors can't usefully recover if state wasn't initialized (failure to call `init_state`) in
-    /// the constructor so this method panics if the state tree if missing
-    fn load_state(&self) -> TokenState {
-        TokenState::load(&self.bs, &self.state_cid).unwrap()
+    /// Flush state and return Cid for root
+    pub fn flush(&mut self) -> Result<Cid> {
+        Ok(self.state.save(&self.bs)?)
     }
 
-    /// Helper function that saves the state tree and returns the CID of the root
-    fn save_state(&mut self, state: &TokenState) -> Result<Cid> {
-        self.state_cid = state.save(&self.bs)?;
-        Ok(self.state_cid)
+    /// Opens an atomic transaction on TokenState which allows a closure to make multiple
+    /// modifications to the state tree.
+    ///
+    /// If the closure returns an error, the transaction is dropped atomically and no change is
+    /// observed on token state.
+    pub fn transaction<F, Res>(&mut self, f: F) -> Result<Res>
```

Comment:
This is interesting but not quite perfect.
1. Ideally don't expose the token state directly to Token users
2. Users can still use the API outside a transaction.

I think this method makes a pretty good _internal_ transaction helper. I.e. call it from within Token methods so they are atomic, but don't expose it publicly. So I reckon make it non-`pub` for this PR, fixing (1).

I'm not sure 2 is a serious problem, but would be _nice_ to address. For application-level transactions, we'd need the callback parameter to be Token or a new TokenTxn trait. If callers strictly used that, it would kinda give (2).